nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -471,23 +481,122 @@ impl DataStore {
+                db::datastore::Disk::LocalStorage(disk) => Some(disk),
+                db::datastore::Disk::Crucible(_) => None,
+            })
+            .collect();
+
+        // Gather constraints for the instance placement, starting with the
+        // arguments to this function.
+        let maybe_must_use_sleds: Option<HashSet<SledUuid>> =
             if let Some(must_select_from) = constraints.must_select_from() {
-                must_select_from.into_iter().collect()
+                Some(must_select_from.into_iter().cloned().collect())
             } else {
-                HashSet::default()
+                None
             };
 
+        // If any local storage disks have been allocated already, then this
+        // constraints VMM placement and where other unallocated local storage
+        // must be.
+        let maybe_must_use_sleds = if !local_storage_disks.is_empty() {
+            // Any local storage disk that was allocated already will have a
+            // sled_id.
+            let local_storage_allocation_sleds: HashSet<SledUuid> =
+                local_storage_disks
+                    .iter()
+                    .filter_map(|disk| {
+                        disk.local_storage_dataset_allocation
+                            .as_ref()
+                            .map(|allocation| allocation.sled_id())
+                    })
+                    .collect();
+
+            if local_storage_allocation_sleds.is_empty() {
+                // None of this instance's local storage disks have been
+                // allocated yet.
+                maybe_must_use_sleds
+            } else {
+                if local_storage_allocation_sleds.len() != 1 {
+                    // It's an error for multiple sleds to host local storage
+                    // disks, that makes no sense!
+                    return Err(SledReservationTransactionError::Connection(
+                        Error::internal_error(&format!(
+                            "local storage disks for instance {instance_id} \
+                            allocated on multiple sleds \
+                            {local_storage_allocation_sleds:?}"
+                        )),
+                    ));
+                }
+
+                // If there's already a list of sleds to use, that must be
+                // filtered further. Otherwise return the one sled that local
+                // storage disks have already been allocated on.
+                match maybe_must_use_sleds {
+                    Some(must_use_sleds) => Some(
+                        must_use_sleds
+                            .into_iter()
+                            .filter(|sled| {
+                                local_storage_allocation_sleds.contains(&sled)
+                            })
+                            .collect(),
+                    ),
+
+                    None => Some(local_storage_allocation_sleds),
+                }
```

Comment:
I think I may have gotten confused. I expected this to basically say "does `maybe_must_use_sleds` contain the (exactly one) item from `local_storage_allocation_sleds`--if it doesn't, that's going to be an error.